File: ios/Classes/Location/CustomerIOLocation.swift
import CioInternalCommon
import CioLocation
import CoreLocation
import Flutter
import Foundation

public class CustomerIOLocation: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private let logger: Logger = DIGraphShared.shared.logger

    public static func register(with _: FlutterPluginRegistrar) {}

    init(with registrar: FlutterPluginRegistrar) {
        super.init()

        methodChannel = FlutterMethodChannel(name: "customer_io_location", binaryMessenger: registrar.messenger())

        guard let methodChannel = methodChannel else {
            print("customer_io_location methodChannel is nil")
            return
        }

        registrar.addMethodCallDelegate(self, channel: methodChannel)
    }

    deinit {
        methodChannel?.setMethodCallHandler(nil)
        methodChannel = nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setLastKnownLocation":
            call.nativeMapArgs(result: result, handler: setLastKnownLocation)

        case "requestLocationUpdate":
            call.nativeNoArgs(result: result, handler: requestLocationUpdate)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func setLastKnownLocation(params: [String: AnyHashable]) {
        guard let latitude = params["latitude"] as? Double,
              let longitude = params["longitude"] as? Double
        else {
            logger.error("Latitude and longitude are required for setLastKnownLocation")
            return
        }

        let location = CLLocation(latitude: latitude, longitude: longitude)
        CustomerIO.location.setLastKnownLocation(location)
    }

    private func requestLocationUpdate() {
        CustomerIO.location.requestLocationUpdate()
    }
}

File: android/src/main/kotlin/io/customer/customer_io/location/CustomerIOLocation.kt
package io.customer.customer_io.location

import io.customer.customer_io.bridge.NativeModuleBridge
import io.customer.customer_io.bridge.nativeMapArgs
import io.customer.customer_io.bridge.nativeNoArgs
import io.customer.customer_io.utils.getAs
import io.customer.location.LocationModuleConfig
import io.customer.location.LocationTrackingMode
import io.customer.location.ModuleLocation
import io.customer.sdk.CustomerIOBuilder
import io.customer.sdk.core.di.SDKComponent
import io.customer.sdk.core.util.Logger
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter module implementation for location module in native SDKs. All functionality
 * linked with the module should be placed here.
 */
internal class CustomerIOLocation(
    pluginBinding: FlutterPlugin.FlutterPluginBinding,
) : NativeModuleBridge, MethodChannel.MethodCallHandler {
    override val moduleName: String = "Location"
    override val flutterCommunicationChannel: MethodChannel =
        MethodChannel(pluginBinding.binaryMessenger, "customer_io_location")
    private val logger: Logger = SDKComponent.logger

    private fun getLocationServices() = runCatching {
        ModuleLocation.instance().locationServices
    }.onFailure {
        logger.error("Location module is not initialized. Ensure location config is provided during SDK initialization.")
    }.getOrNull()

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setLastKnownLocation" -> call.nativeMapArgs(result, ::setLastKnownLocation)
            "requestLocationUpdate" -> call.nativeNoArgs(result, ::requestLocationUpdate)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun setLastKnownLocation(params: Map<String, Any>) {
        val latitude = params.getAs<Double>("latitude")
        val longitude = params.getAs<Double>("longitude")

        if (latitude == null || longitude == null) {
            logger.error("Latitude and longitude are required for setLastKnownLocation")
            return
        }

        getLocationServices()?.setLastKnownLocation(latitude, longitude)
    }

    private fun requestLocationUpdate() {
        getLocationServices()?.requestLocationUpdate()
    }

    override fun configureModule(
        builder: CustomerIOBuilder,
        config: Map<String, Any>
    ) {
        val trackingModeValue = config.getAs<String>("trackingMode")
        val trackingMode = trackingModeValue?.let { value ->
            runCatching { enumValueOf<LocationTrackingMode>(value) }.getOrNull()
        } ?: LocationTrackingMode.MANUAL

        val module = ModuleLocation(
            LocationModuleConfig.Builder()
                .setLocationTrackingMode(trackingMode)
                .build()
        )
        builder.addCustomerIOModule(module)
    }
}
